nexusparser/tools/dataconverter/convert.py

```diff
@@ -0,0 +1,157 @@
+"""This script runs the conversion routine using a selected reader and write out a Nexus file."""
+
+import glob
+import importlib.machinery
+import importlib.util
+import json
+import logging
+import os
+import sys
+from typing import List, Tuple, Dict
+import xml.etree.ElementTree as ET
+
+import click
+
+from nexusparser.tools.dataconverter.writer import Writer
+
+logger = logging.getLogger(__name__)  # pylint: disable=C0103
+logger.setLevel(logging.INFO)
+logger.addHandler(logging.StreamHandler(sys.stdout))
+
+
+# Nexus related functions to be exported in a common place for all tools
+def convert_nexus_to_caps(nexus):
+    """Helper function to convert a Nexus class from <NxClass> to <CLASS>."""
+    return nexus[2:].upper()
+
+
+def convert_nexus_to_suggested_name(nexus):
+    """Helper function to suggest a name for a group from its Nexus class."""
+    return nexus[2:]
+
+
+# Common XML function
+def remove_namespace_from_tag(tag):
+    """Helper function to remove the namespace from an XML tag."""
+    return tag.split("}")[-1]
+
+
+# Helper functions for the convert script
+def get_first_group(root):
+    """Helper function to get the actual first group element from the NXDL"""
+    for child in root:
+        if remove_namespace_from_tag(child.tag) == "group":
+            return child
+    return root
+
+
+def generate_template_from_nxdl(root, path, template):
+    """Helper function to generate a template dictionary for given NXDL"""
+    tag = remove_namespace_from_tag(root.tag)
+
+    if tag == "doc":
+        return
+
+    suffix = ""
+    if "name" in root.attrib:
+        suffix = root.attrib['name']
+    elif "type" in root.attrib:
+        nexus_class = convert_nexus_to_caps(root.attrib['type'])
+        hdf5name = f"[{convert_nexus_to_suggested_name(root.attrib['type'])}]"
+        suffix = f"{nexus_class}{hdf5name}"
+
+    if tag == "attribute":
+        suffix = f"@{suffix}"
+
+    path = path + "/" + suffix
+
+    # Only add fields or attributes to the dictionary
+    if tag in ("field", "attribute"):
+        template[path] = None
+
+    # Only add units if it is a field and the the units are not set to NX_UNITLESS
+    if tag == "field" and "units" in root.attrib.keys() and root.attrib["units"] != "NX_UNITLESS":
+        template[f"{path}/@units"] = None
+
+    for child in root:
+        generate_template_from_nxdl(child, path, template)
+
+
+def get_reader(reader_name):
+    """Helper function to get the reader object from it's given name"""
+    path_prefix = f"{os.path.dirname(__file__)}/" if os.path.dirname(__file__) else ""
+    path = f"{path_prefix}readers/{reader_name}_reader.py"
+    spec = importlib.util.spec_from_file_location(f"{reader_name}_reader.py", path)
+    module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(module)
+    return module.READER
```

Comment:
I chose this way because then I only instantiate one reader. Usually, all plugins need to be available. Here, we only work with a single reader during an execution. If I go with the fancy way you mention, then I'll have to instantiate all the readers so that they are registered. With this approach only the needed reader is loaded and instantiated. This just seems to be more efficient. If you suggest a way to only load one of them in or you think it's no worries to instantiate all of them and it will make code readability better, I can switch to the fancy registration.

As for having a dictionary, I kinda do that already with the names with the get_names_of_all_readers function. It just populates on every call instead of storing them forever. I only need the list once for the click.choice for my --reader. And having a list would still require me to instantiate the classes. And that means the approach would be similar.

I really don't want users to provide class names. Once, these readers are implemented, the users would just specify what domain specific reader they are converting from. Rather than what class they are using. And the name of the reader seems more intuitive.